sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -1031,78 +1014,163 @@ ProgramManager::ProgramManager() {
       throw runtime_error(std::string("read from ") + SpvFile +
                               std::string(" failed"),
                           PI_ERROR_INVALID_VALUE);
-    auto ImgPtr =
+    // No need for a mutex here since all access to these private fields is
+    // blocked until the construction of the ProgramManager singleton is
+    // finished.
+    m_SpvFileImage =
         make_unique_ptr<DynRTDeviceBinaryImage>(std::move(Data), Size);
 
     if (DbgProgMgr > 0) {
       std::cerr << "loaded device image binary from " << SpvFile << "\n";
-      std::cerr << "format: " << getFormatStr(ImgPtr->getFormat()) << "\n";
+      std::cerr << "format: " << getFormatStr(m_SpvFileImage->getFormat())
+                << "\n";
     }
-    // No need for a mutex here since all access to these private fields is
-    // blocked until the construction of the ProgramManager singleton is
-    // finished.
-    m_DeviceImages[SpvFileKSId].reset(
-        new std::vector<RTDeviceBinaryImageUPtr>());
-    m_DeviceImages[SpvFileKSId]->push_back(std::move(ImgPtr));
   }
 }
 
+void CheckJITCompilationForImage(const RTDeviceBinaryImage *const &Image,
+                                 bool JITCompilationIsRequired) {
+  if (!JITCompilationIsRequired)
+    return;
+  // If the image is already compiled with AOT, throw an exception.
+  const pi_device_binary_struct &RawImg = Image->getRawData();
+  if ((strcmp(RawImg.DeviceTargetSpec,
+              __SYCL_PI_DEVICE_BINARY_TARGET_SPIRV64_X86_64) == 0) ||
+      (strcmp(RawImg.DeviceTargetSpec,
+              __SYCL_PI_DEVICE_BINARY_TARGET_SPIRV64_GEN) == 0) ||
+      (strcmp(RawImg.DeviceTargetSpec,
+              __SYCL_PI_DEVICE_BINARY_TARGET_SPIRV64_FPGA) == 0)) {
+    throw feature_not_supported("Recompiling AOT image is not supported",
+                                PI_ERROR_INVALID_OPERATION);
+  }
+}
+
+template <typename StorageKey>
+RTDeviceBinaryImage *getBinImageFromMultiMap(
+    const std::unordered_multimap<StorageKey, RTDeviceBinaryImage *> &ImagesSet,
+    const StorageKey &Key, const context &Context, const device &Device) {
+  auto [ItBegin, ItEnd] = ImagesSet.equal_range(Key);
+  if (ItBegin == ItEnd)
+    return nullptr;
+
+  std::vector<pi_device_binary> RawImgs(std::distance(ItBegin, ItEnd));
+  auto It = ItBegin;
+  for (unsigned I = 0; It != ItEnd; ++It, ++I)
+    RawImgs[I] = const_cast<pi_device_binary>(&It->second->getRawData());
```

Comment:
`const_cast`? Can we leave without it?

Reply:
this const cast is not introduced by my change, just moved.
The method that is used here has const return value and const itself:
_const pi_device_binary_struct &getRawData()_. Then we are getting pointer to that value that has type _const pi_device_binary_struct*_.

piextDeviceSelectBinary where array is passed to accepts non const data that leads to the necessity of this cast. We could remove const from getRawData return value but I prefer to do cast explicitly only where it is needed and keep it const.
What do you think?